src/lib/mutate-types.ts

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023-Present The Pepr Authors
+
+import { KubernetesObject } from "kubernetes-fluent-client";
+import { Logger } from "pino";
+import { GroupVersionKind, GenericClass } from "kubernetes-fluent-client";
+import { clone, mergeDeepRight } from "ramda";
+
+// Basic operation type for mutation operations
+export enum Operation {
+  CREATE = "CREATE",
+  UPDATE = "UPDATE",
+  DELETE = "DELETE",
+  CONNECT = "CONNECT",
+}
+
+// AdmissionRequest interface for handling admission requests in the mutation context
+export interface AdmissionRequest<T = KubernetesObject> {
+  readonly uid: string;
+  readonly kind: GroupVersionKind;
+  readonly resource: GroupVersionResource;
+  readonly subResource?: string;
+  readonly requestKind?: GroupVersionKind;
+  readonly requestResource?: GroupVersionResource;
+  readonly requestSubResource?: string;
+  readonly name: string;
+  readonly namespace?: string;
+  readonly operation: Operation;
+  readonly userInfo: {
+    username?: string;
+    uid?: string;
+    groups?: string[];
+    extra?: { [key: string]: string[] };
+  };
+  readonly object: T;
+  readonly oldObject?: T;
+  readonly dryRun?: boolean;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  readonly options?: any;
+}
+
+// DeepPartial utility type for deep optional properties
+export type DeepPartial<T> = {
+  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
+};
+
+// MutateAction type for handling mutation callbacks
+export type MutateAction<T extends GenericClass, K extends KubernetesObject = InstanceType<T>> = (
+  req: PeprMutateRequest<K>,
+  logger?: Logger,
+) => Promise<void> | void | Promise<PeprMutateRequest<K>> | PeprMutateRequest<K>;
+
+// PeprMutateRequest class for mutation request handling
+export class PeprMutateRequest<T extends KubernetesObject> {
```

Comment:
See other comment about leaving these in `mutate-request.ts`.

Reply:
Fixed with the changes you suggested